File: benchmarks-ocaml/scott_quicksort.ml
type 'a scott_list = { unlist : 'r. 'r -> ('a -> 'a scott_list -> 'r) -> 'r }

let scott_nil = { unlist = (fun n c -> n) }

let scott_cons h t = { unlist = (fun n c -> c h t) }

let rec scott_filter f list =
  list.unlist scott_nil (fun x xs ->
      if f x then scott_cons x (scott_filter f xs) else scott_filter f xs)

let rec scott_append xs ys =
  xs.unlist ys (fun x xss -> scott_cons x (scott_append xss ys))

let rec scott_quicksort list =
  list.unlist scott_nil (fun x xs ->
      scott_append
        (scott_quicksort (scott_filter (fun y -> y < x) xs))
        (scott_cons x (scott_quicksort (scott_filter (fun z -> z >= x) xs))))

let rec scott_sum_list list = list.unlist 0 (fun x xs -> x + scott_sum_list xs)

let generate_list n =
  let rec go i acc = if i < n then go (i + 1) (scott_cons i acc) else acc in
  go 0 scott_nil

let () =
  print_int (scott_sum_list (scott_quicksort (generate_list 1000)));
  print_newline ()

File: benchmarks-haskell/scott-quicksort.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}

{-# HLINT ignore "Use const" #-}
{-# HLINT ignore "Avoid lambda using `infix`" #-}
{-# HLINT ignore "Redundant bracket" #-}

newtype ScottList a = ScottList (forall r. r -> (a -> ScottList a -> r) -> r)

unlist :: ScottList a -> forall r. r -> (a -> ScottList a -> r) -> r
unlist (ScottList f) = f

scottNil :: ScottList a
scottNil = ScottList (\n c -> n)

scottCons :: a -> ScottList a -> ScottList a
scottCons h t = ScottList (\n c -> c h t)

scottFilter :: (Int -> Bool) -> ScottList Int -> ScottList Int
scottFilter f list =
  (unlist list)
    scottNil
    ( \x xs ->
        if f x
          then scottCons x (scottFilter f xs)
          else scottFilter f xs
    )

scottAppend :: ScottList Int -> ScottList Int -> ScottList Int
scottAppend xs ys = (unlist xs) ys (\x xss -> scottCons x (scottAppend xss ys))

scottQuicksort :: ScottList Int -> ScottList Int
scottQuicksort list =
  (unlist list)
    scottNil
    ( \x xs ->
        scottAppend
          (scottQuicksort (scottFilter (\y -> y < x) xs))
          (scottCons x (scottQuicksort (scottFilter (\z -> z >= x) xs)))
    )

scottSumList :: ScottList Int -> Int
scottSumList list = (unlist list) 0 (\x xs -> x + scottSumList xs)

generateList :: Int -> ScottList Int
generateList n = go 0 scottNil
  where
    go i acc
      | i < n = go (i + 1) (scottCons i acc)
      | otherwise = acc

main :: IO ()
main =
  print $ scottSumList $ scottQuicksort $ generateList 1000
